_posts/2020-02-17-eda-architecture.md

```diff
@@ -0,0 +1,43 @@
+---
+layout: post
+title:  "An event driven architecture"
+excerpt: "Why and How we ditched our old architecture based on synchronous REST services for a completely asynchronous event driven architecture."
+date:   2020-02-24 12:53:00 +0100
+categories: architecture
+tags: services event bus architecture
+---
+
+## Introduction
+
+At [Allo-Media](https://www.allo-media.net/en/), like many other businesses, our value chain looks like a pipeline: we collect conversations (mainly through phone) sent to us by our customers, we transcribe them, we tag the transcripts with named entities, we anonymize both the transcript and the audio, then we qualify the content with semantic tags, and finally we index them and provide a UI and API to consult, search, analyze the conversations. All those steps are completed automatically by NLP and AI algorithms.
+
+Such pipelines are well suited for service based architectures. If you need to add a new feature, you introduce a new service into the pipeline.
+
+Our first take at it was based on REST services. Unfortunately, that approach had many drawbacks:
+* it introduces strong coupling between components, as each service has to know about the other related services, their addresses, their purposes, their APIs…;
+* load balancing requires ad-hoc solutions;
+* high availability is tricky because of the synchronous communication: if the requested service is down, the caller has to implement complex "retry later" strategies or give up! And so on for each service.
+* upgrading or adding new services is a lot of work as it impacts other services and requires careful coordinated releases. Plus, you have to provide them with IP and DNS addresses.
+
+So one year later, as our activity grew and development accelerated, we quickly realized that we needed:
+
+* maximum service decoupling;
+* easy distribution;
+* no brainer load balancing;
```

Comment:
But here the dash is required for no-brainer: https://www.merriam-webster.com/dictionary/no-brainer